### ISSUE TITLE
WM_CHAR input message repeat instantly without delay after first character

### ISSUE THREAD

**Issue by JX-Master** [ocornut/imgui#1808]
----

**Version/Branch of Dear ImGui:**

v1.60

**Back-end/Renderer/OS:**
- Windows 10 Pro build 16299.
- System Language: English.
- Render: DirectX 11.0
- IDE: Visual Studio 2017 Exterprise. Visual C++ V141 Compiler.
- Utilize my own rendering implementation modified from Dx11 example shipped with v1.60.

**Hardware Info:**
- CPU : AMD Ryzen 1800x Eight-core Processor.
- GPU: GTX1080ti
- Memory: 24GB

**My Issue/Question:** 
The Input text repeats instantly without any delay even if I set **io.KeyRepeatDelay correctly using default value 0.25f, and the io.DeltaTime is updated correctly, wiggling around 0.016 since V-Sync is on**.

I'm using default English Keyboard and English IME (United States).

This problem only happens when inputting characters like numbers, letters and punctuations. **Backspaces work out fine with the io.KeyRepeatDelay applied correctly.**

The Dx11 example shipped with v1.60 doesn't have this problem, so I think there must be something missing when migrating the code from example file to my application framework.

I checked the WM_CHAR event when I tried to figure out the problem. It seems that in the DX11 example code, the second WM_CHAR event delayed a litter bit after the first WM_CHAR event is triggered. **This delay is not being controlled by io.KeyRepeatDelay.** While in my code, the second WM_CHAR event triggers instantly without any delay, so when I press a key of character, it instantly spawns a series of WM_CHAR events. The number of events triggered is based on how fast I release the key, ranging from 1 to infinite. :(

I didn't do much change to imgui_impl_dx11.cpp file, only intergrated the ImGui_ImplWin32_WndProcHandler(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam) function directly to my WinProc function because referencing another WndProc function in my code will somehow cause the Window Registration failed.

**Standalone, minimal, complete and verifiable example:** 
Here is my WndProc function. It is not standalone actually, but it's almost like what it writes in imgui_impl_dx11.cpp. Any other functions related to ImGui is directly moved from imgui_impl_dx11.cpp without any change except modifying the ID3D11XX interface to my own resources.
```
LRESULT CALLBACK WndProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
{
	static TWorld world(*pWorld);
	static auto pWindowInfo = world.FetchWorldState<SWindowInfo>();
	LRESULT returnValue = 0;
	switch (message)
    {
	case WM_SIZE:
		// Do some common DX11 resize thing.
		break;
	case WM_ENTERSIZEMOVE:
		// The Application will not update when paused.
		pWindowInfo->bAppPaused = true;
		pWindowInfo->bResizing = true;
		// Time is the Timer I use to query system performance counter.
		pWindowInfo->Time.Stop();
		break;
	case WM_EXITSIZEMOVE:
		pWindowInfo->bAppPaused = false;
		pWindowInfo->bResizing = false;
		pWindowInfo->Time.Start();
		OnResize(*pWorld);
		break;
    case WM_DESTROY:
        PostQuitMessage(0);
        break;
	case WM_LBUTTONDOWN: case WM_LBUTTONDBLCLK:
	case WM_RBUTTONDOWN: case WM_RBUTTONDBLCLK:
	case WM_MBUTTONDOWN: case WM_MBUTTONDBLCLK:
	{
		int button = 0;
		if (message == WM_LBUTTONDOWN || message == WM_LBUTTONDBLCLK) button = 0;
		if (message == WM_RBUTTONDOWN || message == WM_RBUTTONDBLCLK) button = 1;
		if (message == WM_MBUTTONDOWN || message == WM_MBUTTONDBLCLK) button = 2;
		if (!ImGui::IsAnyMouseDown() && ::GetCapture() == NULL)
			::SetCapture(hWnd);
		ImGuiIO& io = ImGui::GetIO();
		io.MouseDown[button] = true;
		break;
	}
	case WM_LBUTTONUP:
	case WM_RBUTTONUP:
	case WM_MBUTTONUP:
	{
		ImGuiIO& io = ImGui::GetIO();
		int button = 0;
		if (message == WM_LBUTTONUP) button = 0;
		if (message == WM_RBUTTONUP) button = 1;
		if (message == WM_MBUTTONUP) button = 2;
		io.MouseDown[button] = false;
		if (!ImGui::IsAnyMouseDown() && ::GetCapture() == hWnd)
			::ReleaseCapture();
		break;
	}
	case WM_MOUSEWHEEL:
	{
		ImGuiIO& io = ImGui::GetIO();
		io.MouseWheel += GET_WHEEL_DELTA_WPARAM(wParam) > 0 ? +1.0f : -1.0f;
		break;
	}
		
	case WM_MOUSEHWHEEL:
	{
		ImGuiIO& io = ImGui::GetIO();
		io.MouseWheelH += GET_WHEEL_DELTA_WPARAM(wParam) > 0 ? +1.0f : -1.0f;
		break;
	}
	case WM_MOUSEMOVE:
	{
		ImGuiIO& io = ImGui::GetIO();
		io.MousePos.x = (signed short)(lParam);
		io.MousePos.y = (signed short)(lParam >> 16);
		break;
	}
	case WM_KEYDOWN:
	case WM_SYSKEYDOWN:
	{
		ImGuiIO& io = ImGui::GetIO();
		if (wParam < 256)
			io.KeysDown[wParam] = 1;
		if (wParam == VK_OEM_3)	//~ Key
		{
			pWindowInfo->bDisplayDebugWindow = !pWindowInfo->bDisplayDebugWindow;
		}
		if (wParam == VK_ESCAPE)
		{
			PostQuitMessage(0);
		}
		break;
	}
	case WM_KEYUP:
	case WM_SYSKEYUP:
	{
		ImGuiIO& io = ImGui::GetIO();
		if (wParam < 256)
			io.KeysDown[wParam] = 0;
		break;
	}
	case WM_CHAR:
	{
		ImGuiIO& io = ImGui::GetIO();
		// You can also use ToAscii()+GetKeyboardState() to retrieve characters.
		if (wParam > 0 && wParam < 0x10000)
			io.AddInputCharacter((unsigned short)wParam);
		break;
	}
	case WM_SETCURSOR:
	{
		if (LOWORD(lParam) == HTCLIENT)
		{
			ImGuiUpdateMouseCursor();
			returnValue = 1;
		}
		break;
	}
    default:
        return DefWindowProc(hWnd, message, wParam, lParam);
    }
    return returnValue;
}
```

**Screenshots/Video** 
Image:
![problem](https://user-images.githubusercontent.com/25875303/39904403-407d506c-550a-11e8-969c-42d64ea05a55.gif)

**Comment by JX-Master** [comment]
Well it has been 3 days, I have Googled a few times but still couldn't find out the reason. I recreate the whole project, recode everything, and the problem still exists. So I made this MINIMUM standalone runnable project that identifies my problem, which is the "Test" project in my solution file.
It simply initializes the Direct3D 11 environment and draws a simple window.
There is an exe file in "debug" folder in case of any compile error that may occur( It shouldn't although).
[Issue1808.zip](https://github.com/ocornut/imgui/files/1998452/Issue1808.zip)

**Comment by ocornut** [comment]
- Are you sure that a 73 files package constitute a "MINIMUM standalone runnable project" ?

- We have identified at this point that the issue is a Win32 API/system issue and has nothing to do with imgui. The fact that your WM_CHAR handler is called with a fast repeat rate is your issue. 
While I gave additional information on potential improvement for imgui on the inconsistent repeat rate handling, it doesn't directly relate to your issue.
